### PR TITLE
Add Ambient Weather TX-8300 Thermometer

### PR DESCRIPTION
**Pull request by roger-**
Seems to work!

### CODE REVIEW

**Comment by zuckschwerdt** [comment]
Discussion is in #780

File: src/devices/ambient_weather_tx8300.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
+/* Ambient Weather TX-8300 Thermometer
+ * contributed by Roger
+*/
+
+#include <byteswap.h>
```
Comment (zuckschwerdt) [inline src/devices/ambient_weather_tx8300.c]: Put the includes below the protocol info

File: src/devices/ambient_weather_tx8300.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
+/* Ambient Weather TX-8300 Thermometer
```
Comment (zuckschwerdt) [inline src/devices/ambient_weather_tx8300.c]: The device should be named ambientweather_tx8300 -- We'll slowly move to just one word for manufacturer and one for device to simplify parsing in consumers.

File: src/devices/ambient_weather_tx8300.c [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    data_t *data; 
+    int count;
+    int crc;
+    
+	if(bitbuffer->num_rows != 1 || bitbuffer->bits_per_row[0] != BITS_PER_MSG) {
+	    if (debug_output > 1) {
+	        fprintf(stderr, "wrong number of bits");
+	     }
+	        
+		return 0; // Unrecognized data
```
Comment (zuckschwerdt) [inline src/devices/ambient_weather_tx8300.c]: Indentation seems off here

File: src/devices/ambient_weather_tx8300.c [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+Packet format (74 bits):
+   [2 bit preamble] [1 bit start bit] [31 bit payload] [1 bit start bit] [31 bit payload (inverted)] [8 bit CRC]   
+   HH1PPPPP PPPPPPPP PPPPPPPP PPPPPPPP PP1QQQQQ QQQQQQQQ QQQQQQQQ QQQQQQQQ QQCCCCCC CC   
+   
+Preamble format (2 bits):
+    [1 bit (0)] [1 bit rolling count]    
+    0R
+
+Payload format (31 bits):
+   [9 bit unknown (humidity?)] [2 bit channel number] [1 bit negative flag] [7 bit ID]  [12 bit BCD temperature (C, 1 decimal point)]
```
Comment (zuckschwerdt) [inline src/devices/ambient_weather_tx8300.c]: Best to take the payload as 32 bit and mask out the fixed top bit, humidity needs only 7 bits, might be a full byte, but havn't ever seen 9 bits

**Comment by zuckschwerdt** [comment]
Good work. Please change payload to an aligned uint8_t b[] and get all data from that.

**Comment by roger-** [comment]
Thanks. Actually did it that way at first but thought it might be more readable to change it to  use your helper functions (but the endianess issues and extra copied bits defeated the benefits). 

I’ll take a stab at rewriting it though.

File: src/devices/ambient_weather_tx8300.c (hunk 4 above)
Comment (roger-) [reply src/devices/ambient_weather_tx8300.c]: I’m not sure anymore what that “humidity” field is anymore. Could be a combination of battery status, unused humidity or device code or something.

**Comment by zuckschwerdt** [comment]
also `b[0] = (b[0] & 0x7f) | (b[5] & 0x80);` to fix the first bit

**Comment by roger-** [comment]
PR updated.

**Comment by roger-** [comment]
How does it look now? I’ve been running it for a week and all seems fine.

**Comment by zuckschwerdt** [comment]
Oh, I just merged the TFA 30.3211.02 and it seems this is the same device? I'll check and unify both if it turns out to work.

**Comment by zuckschwerdt** [comment]
The Ambient Weather TX-8300 is rebranded as TFA 30.3211.02. See #831.
I unified both PRs and this is now reported as "Ambientweather TX-8300" model.

**Comment by roger-** [comment]
Thanks!

Can you re-add the rolling count?

**Comment by zuckschwerdt** [comment]
Is it then really useful? What are you using that value for?

**Comment by merbanan** [comment]
Sequence number, just let the consumer care about what to use it for.

**Comment by zuckschwerdt** [comment]
It's just a single bit in the preamble, not secured by the checksum. Looks more like an artifact then proper data -- thus my hesitation ;)

**Comment by roger-** [comment]
I’ve been using it to detect (single) dropped packets, but admittedly it isn’t the best way to do that.

**Comment by zuckschwerdt** [comment]
Perhaps just measure the interval between packets, around one minute or such. Some devices vary the time with the selected channel setting. The different times then would be co-prime to ensure the least amount of overlap.

**Comment by zuckschwerdt** [comment]
Sorry for eventual confusion, I've changed the model string to "AmbientWeather-TX8300" to reflect the new direction for model names.

**Comment by klohner** [comment]
FYI: `fprintf(stderr, "H: %02x, F:%02x\n", b[0], b[1] & 0xc0);` in `ambientweather_tx8300.c`
is throwing `H: 0e, F:80` to stderr before each detection I see.

'''
H: 0e, F:80
{"time" : "2018-12-09 19:04:06", "model" : "AmbientWeather-TX8300", "id" : 34, "channel" : 1, "battery" : 2, "temperature_C" : 1.400, "mic" : "CHECKSUM"}
H: 0e, F:80
{"time" : "2018-12-10 13:11:07", "model" : "AmbientWeather-TX8300", "id" : 34, "channel" : 1, "battery" : 2, "temperature_C" : 2.800, "mic" : "CHECKSUM"}
H: 0e, F:80
{"time" : "2018-12-10 13:43:06", "model" : "AmbientWeather-TX8300", "id" : 34, "channel" : 1, "battery" : 2, "temperature_C" : 2.900, "mic" : "CHECKSUM"}
'''

**Comment by zuckschwerdt** [comment]
Yes, that needs to be behind a device->verbose check. Adding it now ;)

**Pull request closed without merging**